File: c/007.c
/*
http://projecteuler.net/problem=7

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?
*/

#include <stdio.h>
#include <math.h>

const int nthPrime = 10001;

int isPrime(int num, int *primes, int indexOfLastPrime) {
	int i;
	double sqrtNum = sqrt(num);
	for (i = 0; i < (indexOfLastPrime + 1); i++) {
		if (primes[i] > sqrtNum) {
			break;
		}
		if (num % primes[i] == 0) {
			return 0;
		}
	}
	return 1;
}

int main()
{
	int primes[nthPrime];
	primes[0] = 2;

	const int nthPrimeIndex = nthPrime - 1;
	int indexOfLastPrime = 0;

	int i;
	for (i = 3; 1 == 1; i++) {
		if (isPrime(i, primes, indexOfLastPrime) == 1) {
			indexOfLastPrime += 1;
			primes[indexOfLastPrime] = i;
		}
		if (indexOfLastPrime == nthPrimeIndex) {
			break;
		}
	}

	printf("The %d prime is %d. [C]\n", nthPrime, primes[nthPrimeIndex]);

    return 0;
}

File: haskell/007.hs
{-
http://projecteuler.net/problem=7

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?
-}

euler007 :: Int
euler007 = nthPrime 10001

-- Determine if the given number is a prime or not.
isPrime :: Int -> Bool
isPrime 2 = True
isPrime x = null [ y | y <- [3,5..((ceiling . sqrt . fromIntegral) x)], x `mod` y == 0]


-- Find the Nth prime number.
-- Hardcode 2 as the first prime, only check odd numbers after that.
nthPrime :: Int -> Int
nthPrime n = last (take n (2:[x | x <- [3,5..], isPrime x]))